R-package/tests/testthat/test_basic.R

```diff
@@ -1247,6 +1247,38 @@ test_that("lgb.train() supports non-ASCII feature names", {
   }
 })
 
+test_that("lgb.train() works with integer, double, and numeric data", {
+  data(mtcars)
+  X <- as.matrix(mtcars[, -1L])
+  y <- mtcars[, 1L, drop = TRUE]
+  expected_mae <- 4.263667
+  for (data_mode in c("numeric", "double", "integer")) {
+    mode(X) <- data_mode
+    nrounds <- 10L
+    bst <- lightgbm(
+      data = X
+      , label = y
+      , params = list(
+        objective = "regression"
+        , min_data = 1L
+        , learning_rate = 0.01
+        , seed = 708L
+      )
+      , nrounds = nrounds
+    )
+
+    # should have trained for 10 iterations and found splits
+    modelDT <- lgb.model.dt.tree(bst)
+    expect_equal(modelDT[, max(tree_index)], nrounds - 1L)
+    expect_gt(nrow(modelDT), nrounds * 3L)
+
+    # should have achieved expected performance
+    preds <- predict(bst, X)
+    mae <- mean(abs(y - preds))
+    expect_true(abs(mae - expected_mae) < TOLERANCE)
+  }
+})
+
```

Comment:
Thanks to @david-cortes for providing the write-up in https://github.com/microsoft/LightGBM/issues/4216, which included the reproducible example that inspired this unit test.

I tested this unit test tonight on my Mac, with R 3.6 and `{lightgbm}` 3.2.1 installed from CRAN, and confirmed that it failed in exactly the way described in #4216. So I'm confident that this test passing means the changes in this PR fix #4216.